control/grpc.py

```diff
@@ -236,42 +260,92 @@ def __init__(self, config: GatewayConfig, gateway_state: GatewayStateHandler, rp
         self.subsys_max_ns = {}
         self.host_info = SubsystemHostAuth()
 
-    def create_host_psk_file(self, subsysnqn : str, hostnqn : str, psk_value : str) -> str:
+    def create_host_key_file(self, key_type : str, subsysnqn : str, hostnqn : str, key_value : str) -> str:
         assert subsysnqn, "Subsystem NQN can't be empty"
         assert hostnqn, "Host NQN can't be empty"
-        assert psk_value, "PSK value can't be empty"
+        assert key_value, "Key value can't be empty"
 
-        psk_dir = f"/tmp/psk/{subsysnqn}"
-        psk_file = f"{psk_dir}/{hostnqn}"
+        key_dir = f"/tmp/{key_type}/{subsysnqn}"
+        key_file = f"{key_dir}/{hostnqn}"
         try:
-            os.makedirs(psk_dir, 0o755, True)
+            os.makedirs(key_dir, 0o755, True)
         except Exception:
-            self.logger.exception(f"Error creating directory {psk_dir}")
+            self.logger.exception(f"Error creating directory {key_dir}")
             return None
         try:
-            with open(psk_file, 'wt') as f:
-                print(psk_value, end="", file=f)
-            os.chmod(psk_file, 0o600)
+            os.remove(key_file)
+        except FileNotFoundError:
+            pass
+        try:
+            with open(key_file, 'wt') as f:
+                print(key_value, end="", file=f)
+            os.chmod(key_file, 0o600)
         except Exception:
-            self.logger.exception(f"Error creating file {psk_file}")
+            self.logger.exception(f"Error creating file {key_file}")
             return None
-        return psk_file
+        return key_file
 
-    def remove_host_psk_file(self, subsysnqn : str, hostnqn : str) -> None:
-        psk_dir = f"/tmp/psk/{subsysnqn}"
-        psk_file = f"{psk_dir}/{hostnqn}"
+    def create_host_psk_file(self, subsysnqn : str, hostnqn : str, key_value : str) -> str:
+        return self.create_host_key_file("psk", subsysnqn, hostnqn, key_value)
+
+    def create_host_dhchap_file(self, subsysnqn : str, hostnqn : str, key_value : str) -> str:
+        return self.create_host_key_file("dhchap", subsysnqn, hostnqn, key_value)
+
+    def create_host_dhchap_controller_file(self, subsysnqn : str, hostnqn : str, key_value : str) -> str:
+        return self.create_host_key_file("dhchap_ctrlr", subsysnqn, hostnqn, key_value)
+
+    def remove_host_key_file(self, key_type : str, subsysnqn : str, hostnqn : str) -> None:
+        key_dir = f"/tmp/{key_type}/{subsysnqn}"
+
+        if not hostnqn:
+            try:
+                shutil.rmtree(key_dir, ignore_errors = True)
```

Comment:
No need of this is you use `tempfile.TemporaryDirectory()`.

Reply:
Again, I need to control the deletion of the directory. When using DHCHAP we can't remove the key files as long we we can still connect using the keys.